pkg/resourcekeeper/statekeep.go

```diff
@@ -67,3 +80,160 @@ func (h *resourceKeeper) StateKeep(ctx context.Context) error {
+	if index == -1 {
+		dotPath := strings.Split(prefix, ".")
+		nestedField, found, err := unstructured.NestedFieldCopy(manifest.UnstructuredContent(), dotPath...)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			return nil, errors.Errorf("the path(%s) was not found ", path)
+		}
+		return nestedField, nil
+	}
+	return nil, nil
+}
+
+// setNestedValueWithSlice set the value to unstructured.Unstructured via the path, like spec.template.spec.containers[0].image
+func setNestedValueWithSlice(manifest *unstructured.Unstructured, path string, value interface{}) (*unstructured.Unstructured, error) {
+	raw := manifest.DeepCopy()
+	var pathSlice []string
+	var indexSlice []int
+	var valueSlice []interface{}
+setLabel:
+	prefix, index, suffix := splitPathWithIndex(path)
+	// 是 slice
+	if index != -1 {
+		dotPath := strings.Split(prefix, ".")
+		nestedSlice, found, err := unstructured.NestedSlice(manifest.UnstructuredContent(), dotPath...)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			return nil, errors.Errorf("the path(%s) was not found ", path)
+		}
+		manifest.Object = nestedSlice[index].(map[string]interface{})
+		pathSlice = append(pathSlice, prefix)
+		indexSlice = append(indexSlice, index)
+		valueSlice = append(valueSlice, nestedSlice)
+		path = suffix
+		goto setLabel
+	}
+	// 不是 slice
+	if index == -1 {
+		pathSlice = append(pathSlice, prefix)
+		indexSlice = append(indexSlice, index)
+		valueSlice = append(valueSlice, nil)
+	}
+
+	// set area
+	resUnstructured := map[string]interface{}{}
+	for i := len(pathSlice) - 1; i >= 0; i-- {
+		// 是 slice
+		if indexSlice[i] != -1 {
+			split := strings.Split(pathSlice[i], ".")
+			err := unstructured.SetNestedSlice(raw.UnstructuredContent(), valueSlice[i].([]interface{}), split...)
+			if err != nil {
+				return nil, err
+			}
+			resUnstructured = raw.Object
+		}
+		// 不是 slice
+		if indexSlice[i] == -1 {
+			split := strings.Split(pathSlice[i], ".")
+			if i-1 >= 0 {
+				temp := valueSlice[i-1].([]interface{})[indexSlice[i-1]].(map[string]interface{})
+				err := unstructured.SetNestedField(temp, value, split...)
+				if err != nil {
+					return nil, err
+				}
+				resUnstructured = temp
+			} else {
+				err := unstructured.SetNestedField(raw.UnstructuredContent(), value, split...)
+				if err != nil {
+					return nil, err
+				}
+				resUnstructured = raw.Object
+			}
+		}
+	}
+	toUnstructured, err := toUnstructured(resUnstructured)
+	if err != nil {
+		return nil, err
+	}
+	return toUnstructured, nil
+}
+
+func splitPathWithIndex(path string) (prefix string, index int, suffix string) {
+	if strings.Contains(path, "[") {
+		prefixSplit := strings.SplitN(path, "[", 2)
+		prefix = prefixSplit[0]
+		suffixSplit := strings.SplitN(prefixSplit[1], "].", 2)
+		index = cast.ToInt(suffixSplit[0])
+		suffix = suffixSplit[1]
+		return prefix, index, suffix
+	}
+	return path, -1, ""
+}
+
+func toUnstructured(in map[string]interface{}) (*unstructured.Unstructured, error) {
```

Comment:
Maybe reuse this function: `Object2Unstructured`

Or write something similar in the common library.

Reply:
I'll use object2unstructured, which I haven't found before